File: backend/crud/messages.py
"""
CRUD operations for Message entities.
"""

import json
from datetime import datetime, timezone
from typing import List

import models
import schemas
from database import retry_on_db_lock, serialized_write
from domain.value_objects.enums import MessageRole, ParticipantType
from sqlalchemy import delete
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def create_message(
    db: AsyncSession, room_id: int, message: schemas.MessageCreate, update_room_activity: bool = True
) -> models.Message:
    """
    Create a new message in the database.

    Args:
        db: Database session
        room_id: Room ID
        message: Message to create
        update_room_activity: Whether to update room's last_activity_at (default: True)

    Returns:
        Created message
    """
    # Serialize anthropic_calls to JSON if present
    anthropic_calls_json = None
    if message.anthropic_calls:
        anthropic_calls_json = json.dumps(message.anthropic_calls)

    db_message = models.Message(
        room_id=room_id,
        agent_id=message.agent_id,
        content=message.content,
        role=message.role,
        participant_type=message.participant_type,
        participant_name=message.participant_name,
        thinking=message.thinking,
        anthropic_calls=anthropic_calls_json,
        image_data=message.image_data,
        image_media_type=message.image_media_type,
        chat_session_id=message.chat_session_id,
    )
    db.add(db_message)

    # Update room's last_activity_at if requested (atomic with message creation)
    if update_room_activity:
        room = await db.get(models.Room, room_id)
        if room:
            room.last_activity_at = datetime.now(timezone.utc)

    async with serialized_write():
        await db.commit()
    await db.refresh(db_message)

    # Invalidate message cache for this room
    from infrastructure.cache import get_cache, room_messages_key

    cache = get_cache()
    # Invalidate all message-related cache entries for this room
    cache.invalidate_pattern(room_messages_key(room_id))

    return db_message


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def create_system_message(
    db: AsyncSession, room_id: int, content: str, update_room_activity: bool = False
) -> models.Message:
    """
    Create a system message (e.g., "invited [agent_name]").

    Args:
        db: Database session
        room_id: Room ID
        content: Message content
        update_room_activity: Whether to update room's last_activity_at (default: False for system messages)

    Returns:
        Created system message
    """
    db_message = models.Message(
        room_id=room_id,
        agent_id=None,
        content=content,
        role=MessageRole.ASSISTANT,
        participant_type=ParticipantType.SYSTEM,
        participant_name=None,
        thinking=None,
    )
    db.add(db_message)

    # Update room's last_activity_at if requested
    if update_room_activity:
        room = await db.get(models.Room, room_id)
        if room:
            room.last_activity_at = datetime.now(timezone.utc)

    async with serialized_write():
        await db.commit()
    await db.refresh(db_message)

    # Invalidate message cache for this room
    from infrastructure.cache import get_cache, room_messages_key

    cache = get_cache()
    cache.invalidate_pattern(room_messages_key(room_id))

    return db_message


async def get_messages(db: AsyncSession, room_id: int) -> List[models.Message]:
    """Get all messages in a room."""
    result = await db.execute(
        select(models.Message)
        .options(selectinload(models.Message.agent))
        .where(models.Message.room_id == room_id)
        .order_by(models.Message.timestamp)
    )
    return result.scalars().all()


async def get_messages_since(
    db: AsyncSession, room_id: int, since_id: int = None, limit: int = 100
) -> List[models.Message]:
    """
    Get messages in a room since a specific message ID.
    Used for polling to fetch only new messages.

    Args:
        db: Database session
        room_id: Room ID
        since_id: Only return messages with ID greater than this (optional)
        limit: Maximum number of messages to return (default: 100, max: 1000)

    Returns:
        List of messages ordered by timestamp (limited to prevent memory issues)
    """
    # Cap limit at 1000 to prevent excessive memory usage
    limit = min(limit, 1000)

    query = select(models.Message).options(selectinload(models.Message.agent)).where(models.Message.room_id == room_id)

    if since_id is not None:
        query = query.where(models.Message.id > since_id)

    query = query.order_by(models.Message.timestamp).limit(limit)

    result = await db.execute(query)
    return result.scalars().all()


async def get_recent_messages(db: AsyncSession, room_id: int, limit: int = 200) -> List[models.Message]:
    """
    Get the most recent messages in a room ordered by timestamp ascending.

    Args:
        db: Database session
        room_id: Room ID
        limit: Maximum number of messages to return (default: 200)

    Returns:
        List of recent messages ordered by timestamp
    """
    query = (
        select(models.Message)
        .options(selectinload(models.Message.agent))
        .where(models.Message.room_id == room_id)
        .order_by(models.Message.timestamp.desc())
        .limit(limit)
    )

    result = await db.execute(query)
    # Reverse to preserve chronological order for context building
    return list(reversed(result.scalars().all()))


async def get_messages_after_agent_response(
    db: AsyncSession, room_id: int, agent_id: int, limit: int = 200
) -> List[models.Message]:
    """
    Get messages posted after the specified agent's last response.

    Falls back to messages after the agent's invitation time when they haven't responded yet.

    Args:
        db: Database session
        room_id: Room ID
        agent_id: Agent ID to filter after
        limit: Maximum number of messages to return

    Returns:
        Chronologically ordered messages after the agent's last response or invitation
    """
    # Find the agent's last message ID (if any)
    last_agent_message_id = await _get_last_agent_message_id(db, room_id, agent_id)

    query = select(models.Message).options(selectinload(models.Message.agent)).where(models.Message.room_id == room_id)

    if last_agent_message_id is not None:
        # Agent has responded before - show messages after their last response
        query = query.where(models.Message.id > last_agent_message_id)
    else:
        # Agent hasn't responded yet - check for invitation timestamp
        joined_at = await _get_agent_joined_at(db, room_id, agent_id)
        if joined_at is not None:
            # Show messages from invitation onwards
            query = query.where(models.Message.timestamp >= joined_at)
        # If no joined_at (old agents), show all recent messages (current behavior)

    query = query.order_by(models.Message.timestamp.desc()).limit(limit)

    result = await db.execute(query)
    return list(reversed(result.scalars().all()))


async def _get_last_agent_message_id(db: AsyncSession, room_id: int, agent_id: int) -> int | None:
    """Get the ID of the most recent message from the given agent in the room."""
    result = await db.execute(
        select(models.Message.id)
        .where(models.Message.room_id == room_id, models.Message.agent_id == agent_id)
        .order_by(models.Message.id.desc())
        .limit(1)
    )
    return result.scalar_one_or_none()


async def _get_agent_joined_at(db: AsyncSession, room_id: int, agent_id: int):
    """Get the timestamp when the agent was added to the room."""
    from sqlalchemy import and_

    result = await db.execute(
        select(models.room_agents.c.joined_at).where(
            and_(models.room_agents.c.room_id == room_id, models.room_agents.c.agent_id == agent_id)
        )
    )
    return result.scalar_one_or_none()


async def get_chat_session_messages(
    db: AsyncSession, room_id: int, chat_session_id: int, limit: int = 100
) -> List[models.Message]:
    """
    Get messages for a specific chat session.

    Used for building conversation context in chat mode - only includes messages
    from the current chat session.

    Args:
        db: Database session
        room_id: Room ID
        chat_session_id: Chat session ID to filter by
        limit: Maximum number of messages to return

    Returns:
        List of messages ordered by timestamp
    """
    query = (
        select(models.Message)
        .options(selectinload(models.Message.agent))
        .where(models.Message.room_id == room_id)
        .where(models.Message.chat_session_id == chat_session_id)
        .order_by(models.Message.timestamp.desc())
        .limit(limit)
    )

    result = await db.execute(query)
    return list(reversed(result.scalars().all()))


async def get_messages_excluding_chat(db: AsyncSession, room_id: int, limit: int = 200) -> List[models.Message]:
    """
    Get messages excluding chat mode messages (where chat_session_id is NULL).

    Used for building game mode context - excludes chat mode conversations
    to keep game context separate.

    Args:
        db: Database session
        room_id: Room ID
        limit: Maximum number of messages to return

    Returns:
        List of game messages ordered by timestamp
    """
    query = (
        select(models.Message)
        .options(selectinload(models.Message.agent))
        .where(models.Message.room_id == room_id)
        .where(models.Message.chat_session_id.is_(None))
        .order_by(models.Message.timestamp.desc())
        .limit(limit)
    )

    result = await db.execute(query)
    return list(reversed(result.scalars().all()))


async def delete_room_messages(db: AsyncSession, room_id: int) -> bool:
    """Delete all messages for a specific room using bulk delete."""
    # First verify the room exists
    room_result = await db.execute(select(models.Room).where(models.Room.id == room_id))
    room = room_result.scalar_one_or_none()

    if not room:
        return False  # Room doesn't exist

    # Delete all messages (may be 0 messages, that's OK)
    async with serialized_write():
        await db.execute(delete(models.Message).where(models.Message.room_id == room_id))
        await db.commit()
    return True  # Success - room exists and messages cleared (even if 0)

File: backend/src/crud/messages.ts
/** CRUD operations for Message entities. Synchronous throughout. */

import { and, asc, count, desc, eq, gt, gte, isNull, type SQL } from 'drizzle-orm'
import type { Db } from '../db'
import type { ParticipantType } from '../domain/enums'
import { agents, messages, roomAgents, rooms, type Message, type MessageRole } from '../db/schema'
import { getCache, roomMessagesKey } from '../infrastructure/cache'

export type { ParticipantType }

// A type alias, not an interface: callers' `GameTime` has no index signature,
// so an interface here would not be assignable.
export type GameTimeSnapshot = { hour: number; minute: number; day: number }

export interface MessageImage {
  data: string
  mediaType: string
}

export interface MessageCreate {
  content: string
  role: MessageRole
  agentId?: number | null
  participantType?: ParticipantType | null
  participantName?: string | null
  thinking?: string | null
  anthropicCalls?: string[] | null
  images?: MessageImage[] | null
  /** Deprecated single-image fields, kept because the column still exists. */
  imageData?: string | null
  imageMediaType?: string | null
  chatSessionId?: number | null
  gameTimeSnapshot?: GameTimeSnapshot | null
}

export interface CreateMessageOptions {
  /** Defaults to true. */
  updateRoomActivity?: boolean
}

// An empty array or object is stored as NULL, not as `"[]"`/`"{}"`.
function jsonOrNull(value: unknown): string | null {
  if (value === null || value === undefined) return null
  if (Array.isArray(value) && value.length === 0) return null
  if (typeof value === 'object' && Object.keys(value as object).length === 0) return null
  return JSON.stringify(value)
}

/** One transaction, so a poller cannot see a message whose room looks idle. */
export function createMessage(
  db: Db,
  roomId: number,
  message: MessageCreate,
  { updateRoomActivity = true }: CreateMessageOptions = {},
): Message {
  const row = db.transaction((tx) => {
    const inserted = tx
      .insert(messages)
      .values({
        roomId,
        agentId: message.agentId ?? null,
        content: message.content,
        role: message.role,
        participantType: message.participantType ?? null,
        participantName: message.participantName ?? null,
        thinking: message.thinking ?? null,
        anthropicCalls: jsonOrNull(message.anthropicCalls),
        images: jsonOrNull(
          message.images?.map((img) => ({ data: img.data, media_type: img.mediaType })),
        ),
        imageData: message.imageData ?? null,
        imageMediaType: message.imageMediaType ?? null,
        chatSessionId: message.chatSessionId ?? null,
        gameTimeSnapshot: jsonOrNull(message.gameTimeSnapshot),
        // Explicit rather than the column's CURRENT_TIMESTAMP default, which
        // has second resolution and would sort differently.
        timestamp: new Date(),
      })
      .returning()
      .get()

    if (updateRoomActivity) {
      tx.update(rooms).set({ lastActivityAt: new Date() }).where(eq(rooms.id, roomId)).run()
    }

    return inserted
  })

  // After the commit: the poll loop runs every 2s against a 5s cache, so
  // skipping this makes a turn's replies miss two or three polls.
  getCache().invalidatePattern(roomMessagesKey(roomId))

  return row
}

/** Note the inverted default: bookkeeping must not reorder a room list. */
export function createSystemMessage(
  db: Db,
  roomId: number,
  content: string,
  { updateRoomActivity = false }: CreateMessageOptions = {},
): Message {
  return createMessage(
    db,
    roomId,
    { content, role: 'assistant', agentId: null, participantType: 'system' },
    { updateRoomActivity },
  )
}

function getLastAgentMessageId(db: Db, roomId: number, agentId: number): number | null {
  const row = db
    .select({ id: messages.id })
    .from(messages)
    .where(and(eq(messages.roomId, roomId), eq(messages.agentId, agentId)))
    .orderBy(desc(messages.id))
    .limit(1)
    .get()
  return row?.id ?? null
}

// Null for rows that predate the `joined_at` column.
function getAgentJoinedAt(db: Db, roomId: number, agentId: number): Date | null {
  const row = db
    .select({ joinedAt: roomAgents.joinedAt })
    .from(roomAgents)
    .where(and(eq(roomAgents.roomId, roomId), eq(roomAgents.agentId, agentId)))
    .get()
  return row?.joinedAt ?? null
}

export interface MessageWithAgent extends Message {
  /** Eager-loaded; prompt building reads the agent name off it. */
  agent: typeof agents.$inferSelect | null
}

/**
 * The room's messages with `agent` resolved. The `id` tiebreaker matters:
 * `Date` is millisecond-resolution, so a burst lands on identical timestamps
 * SQLite may return in any order. `newestFirst` sorts descending, limits, then
 * reverses in memory for the *latest* N.
 */
function selectWithAgent(
  db: Db,
  where: SQL,
  { newestFirst, limit }: { newestFirst: boolean; limit?: number },
): MessageWithAgent[] {
  const ordered = db
    .select({ message: messages, agent: agents })
    .from(messages)
    .leftJoin(agents, eq(messages.agentId, agents.id))
    .where(where)
    .orderBy(
      newestFirst ? desc(messages.timestamp) : asc(messages.timestamp),
      newestFirst ? desc(messages.id) : asc(messages.id),
    )

  const rows = limit === undefined ? ordered.all() : ordered.limit(limit).all()
  const mapped = rows.map((r) => ({ ...r.message, agent: r.agent }))
  return newestFirst ? mapped.reverse() : mapped
}

/**
 * How many agent lines a room holds. Counts on `role`, not "has an agent_id":
 * narration output has both, system messages neither. Zero is what "this room
 * has never produced a turn" means — see `routes/game/polling.ts`.
 */
export function countAssistantMessages(db: Db, roomId: number): number {
  return (
    db
      .select({ total: count() })
      .from(messages)
      .where(and(eq(messages.roomId, roomId), eq(messages.role, 'assistant')))
      .get()?.total ?? 0
  )
}

/** Unbounded, for full-history consumers; polling uses {@link getMessagesSince}. */
export function getMessages(db: Db, roomId: number): MessageWithAgent[] {
  return selectWithAgent(db, eq(messages.roomId, roomId), { newestFirst: false })
}

/**
 * The polling read, hit every 2s per open room, so `limit` is clamped rather
 * than rejected. Sorts *ascending* before limiting, unlike its neighbours: a
 * client that fell behind catches up from the oldest unseen.
 */
export function getMessagesSince(
  db: Db,
  roomId: number,
  sinceId: number | null = null,
  limit = 100,
): MessageWithAgent[] {
  const where =
    sinceId === null
      ? eq(messages.roomId, roomId)
      : and(eq(messages.roomId, roomId), gt(messages.id, sinceId))!

  return selectWithAgent(db, where, { newestFirst: false, limit: Math.min(limit, 1000) })
}

/** The newest `limit` messages in a room, returned oldest first. */
export function getRecentMessages(db: Db, roomId: number, limit = 200): MessageWithAgent[] {
  return selectWithAgent(db, eq(messages.roomId, roomId), { newestFirst: true, limit })
}

/**
 * Chat mode and gameplay share a room and a table, and `chat_session_id` is the
 * entire separation: widening this or {@link getMessagesExcludingChat} leaks one
 * mode's transcript into the other's prompt.
 */
export function getChatSessionMessages(
  db: Db,
  roomId: number,
  chatSessionId: number,
  limit = 100,
): MessageWithAgent[] {
  const where = and(eq(messages.roomId, roomId), eq(messages.chatSessionId, chatSessionId))!
  return selectWithAgent(db, where, { newestFirst: true, limit })
}

/** `IS NULL`, not `eq(col, null)` — that renders `= NULL` and matches nothing. */
export function getMessagesExcludingChat(db: Db, roomId: number, limit = 200): MessageWithAgent[] {
  const where = and(eq(messages.roomId, roomId), isNull(messages.chatSessionId))!
  return selectWithAgent(db, where, { newestFirst: true, limit })
}

/**
 * The agent's "what did I miss": everything above its last message id; else
 * everything since its `joined_at`, so a late arrival skips the backlog (an
 * invite has no id, hence the timestamp); else the recent window.
 */
export function getMessagesAfterAgentResponse(
  db: Db,
  roomId: number,
  agentId: number,
  limit = 200,
): MessageWithAgent[] {
  const lastAgentMessageId = getLastAgentMessageId(db, roomId, agentId)

  let where = eq(messages.roomId, roomId)
  if (lastAgentMessageId !== null) {
    where = and(where, gt(messages.id, lastAgentMessageId))!
  } else {
    const joinedAt = getAgentJoinedAt(db, roomId, agentId)
    if (joinedAt !== null) {
      where = and(where, gte(messages.timestamp, joinedAt))!
    }
  }

  return selectWithAgent(db, where, { newestFirst: true, limit })
}

/**
 * The boolean is "no such room" vs "the room now has no messages" — hence the
 * existence check rather than the delete's row count. **No cache invalidation,
 * deliberately:** the caller must sweep, since agent sessions and the client
 * pool have to go down in the same breath.
 */
export function deleteRoomMessages(db: Db, roomId: number): boolean {
  const room = db.select({ id: rooms.id }).from(rooms).where(eq(rooms.id, roomId)).get()
  if (!room) return false

  db.delete(messages).where(eq(messages.roomId, roomId)).run()
  return true
}
